src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -535,6 +554,38 @@ def calculate_dense(self):
     def expectation(self, state, normalize=False):
         return Hamiltonian.expectation(self, state, normalize)
 
+    def expectation_from_samples(self, freq, qubit_map=None):
+        obs = self.terms
+        for term in obs:
+            for factor in term.factors:
+                if isinstance(factor, Z) == False:
+                    raise_error(
+                        NotImplementedError, "Observable is not a Z Pauli string."
+                    )
```

Comment:
I am not sure if this and other symbol manipulations below will fail for more complicated expressions. For example what happens if you have a constant coefficient, eg. `2 * Z(0) * Z(1)` or powers, eg. `Z(0) ** 3` (not very meaningful in terms of physics, but it may appear somehow)? The easiest way to check is to add a few more tests with different hamiltonian forms. 

I remember with other symbolic manipulations we have found several bugs as people were using the code, so these are a bit hard to control.